include/counter.h

```diff
@@ -22,35 +22,119 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 #include <device.h>
+#include <stdbool.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+struct counter_alarm_callback;
+
+/** @brief Alarm callback
+ *
+ * @param dev     Pointer to the device structure for the driver instance.
+ * @param cb      Original structure owning this handler.
+ * @param counter Counter value that triggered the callback.
+ */
+typedef void (*counter_alarm_callback_handler_t)(struct device *dev,
+					 struct counter_alarm_callback *cb,
+					 u32_t counter);
+
+/** @brief Alarm callback structure.
+ *
+ * Used to set an alarm. *
+ * Beware such structure should not be allocated on stack.
+ */
+struct counter_alarm_callback {
+	u8_t channel_id; /*!< Channel ID. Number of channels is driver
+			  * dependent.
+			  */
+
+	bool absolute; /*!< If true ticks are treated as absolute value, else
+			*   it is relative to the counter reading performed
+			*   during the call.
+			*/
+
+	u32_t ticks; /*!< In case of absolute flag is set, maximal value that
+		      *   can be set equals wrap value (counter_get_wrap).
+		      *   Otherwise counter_get_max_relative_alarm() returns
+		      *   maximal value that can be set. If counter is clock
+		      *   driven then ticks can be converted to microseconds
+		      *   (see @ref counter_ticks_to_us). Alternatively,
+		      *   counter implementation may count asynchronous events.
+		      */
+
+	u32_t period; /*!< If non-zero alarm is repeated. If counter is clock
+		       *   driven then period can be converted to microseconds
+		       *   (see @ref counter_ticks_to_us). Alternatively,
+		       *   counter implementation may count asynchronous events.
+		       */
+
+	counter_alarm_callback_handler_t handler; /*!< handler called on alarm
```

Comment:
The handler shouldn't be in the same structure as the alarm parameters. I suggest pulling this out and renaming the struct to counter_alarm_cfg to follow what we did in watchdog.h.

Reply:
But watchdog has handler in parameters (called cfg there): https://github.com/zephyrproject-rtos/zephyr/blob/7fd3eb60d97e645b8ae30e744597496885afada9/include/watchdog.h#L149
What advantage that would have?

Reply:
@MaureenHelm regarding:
> Can we add the concept of multiple channels to the API?
structure counter_alarm_callback has field channel_id. It is intended to use multiple compare channels if present. Channel_id starts from 0. If channel_id is not supported by given hw it will return -ENOTSUP.

Do you have something else in mind?